### PR TITLE
Add maven dependency for JMH framework.

### PR DESCRIPTION
**Pull request by C0rWin**
In order to provide patch for LANG-1110, required dependcy on JMH lib.
Current commit add benchmark profile and ability to run JMH based benchmark by
executing "mvn test -P benchmark" command, moreover it's also possible to
specify exact benchmark name by running "mvn test -P benchmark
-Dbenchmark=benchmark.full.class.name".

### CODE REVIEW

**Comment by C0rWin** [comment]
No Java code was involved in this PR, how that possible coverage reduced?

**Comment by PascalSchumacher** [comment]
@C0rWin The OpenJDK 7 Build did not finish: https://travis-ci.org/apache/commons-lang/jobs/151969189 maybe that caused the change. Or there is a random element in the existing test?

**Comment by C0rWin** [comment]
@PascalSchumacher As far as I know coverage could be reduced only if I've added new code which is not tested or not covered by existing tests.

File: pom.xml [hunk 1]
```diff
@@ -563,6 +578,10 @@
     <commons.site.path>lang</commons.site.path>
     <commons.scmPubUrl>https://svn.apache.org/repos/infra/websites/production/commons/content/proper/commons-lang</commons.scmPubUrl>
     <commons.scmPubCheckoutDirectory>site-content</commons.scmPubCheckoutDirectory>
+
+    <!-- JMH Benchmark related properties, version, target compiler and name of the benchmarking uber jar. -->
```
Comment (PascalSchumacher) [inline pom.xml]: Either the comment is wrong or there are properties missing?

**Comment by kinow** [comment]
Hi @C0rWin 

Some time ago I found this pull request and had to learn what was JMH :-) added a comment on [LANG-1110](https://issues.apache.org/jira/browse/LANG-1110), and today had time to revisit the issue and take a look at the replies.

So [Commons CSV](https://github.com/apache/commons-csv/blob/a7757849d3e9b2a37967ce208ab22f6d14567c98/pom.xml#L389) and [Commons RNG](https://github.com/apache/commons-rng/tree/688f3611131e05a072d998edb93e13492c3ed256/commons-rng-jmh) use JMH too. Both projects apply a similar approach, with the main difference being that Commons RNG - maybe for being a multi-module project - contains a jmh Maven module.

However, the dependencies and plug-in execution are part of a benchmark profile. Following the suggestions in [LANG-1110](https://issues.apache.org/jira/browse/LANG-1110), would be nice if we:

* Used a similar approach, such as profile, class names, etc
* Update [jmh dependencies to the latest](http://search.maven.org/#search%7Cga%7C1%7Cjmh-core)

I wanted to check with you before I opened a separate pull request. Would you be willing to update this pull request again with these changes? If so, just ping me and I'll reply here as soon as possible. Have plenty of time to work on Open Source projects in the next two months :-)

Cheers
Bruno

**Comment by C0rWin** [comment]
Hi Bruno,


On Sun, Feb 12, 2017 at 6:35 AM, Bruno P. Kinoshita <
notifications@github.com> wrote:

> Hi @C0rWin <https://github.com/C0rWin>
>
> Some time ago I found this pull request and had to learn what was JMH :-)
> added a comment on LANG-1110
> <https://issues.apache.org/jira/browse/LANG-1110>, and today had time to
> revisit the issue and take a look at the replies.
>
> So Commons CSV
> <https://github.com/apache/commons-csv/blob/a7757849d3e9b2a37967ce208ab22f6d14567c98/pom.xml#L389>
> and Commons RNG
> <https://github.com/apache/commons-rng/tree/688f3611131e05a072d998edb93e13492c3ed256/commons-rng-jmh>
> use JMH too. Both projects apply a similar approach, with the main
> difference being that Commons RNG - maybe for being a multi-module project
> - contains a jmh Maven module.
>
> However, the dependencies and plug-in execution are part of a benchmark
> profile. Following the suggestions in LANG-1110
> <https://issues.apache.org/jira/browse/LANG-1110>, would be nice if we:
>
>    - Used a similar approach, such as profile, class names, etc
>    - Update jmh dependencies to the latest
>    <http://search.maven.org/#search%7Cga%7C1%7Cjmh-core>
>
> I wanted to check with you before I opened a separate pull request. Would
> you be willing to update this pull request again with these changes? If so,
> just ping me and I'll reply here as soon as possible. Have plenty of time
> to work on Open Source projects in the next two months :-)
>

I can update my pull request to align it with according to your request.

Thanks,
    Artem Barger.

**Comment by PascalSchumacher** [comment]
I guess this is ready to merge?

**Comment by kinow** [comment]
I was waiting for @C0rWin to update the pull request, to have an approach similar to what we have in other commons projects. If he has updated it, then +1 to merging (or maybe check with Emmanuel Bourg if it looks good too?)

**Comment by PascalSchumacher** [comment]
As far as I can tell @C0rWin updated the pull request 9 days ago (after your conversation).

By the way merging this only makes sense if commons lang actually uses JHM. Do you (or somebody else) plan to work on this?

**Comment by kinow** [comment]
Oh, missed that one, will take a look today to compare with [csv].

I've never seen it failing in Jenkins or Travis CI, but I suspect it is run manually, whenever someone needs to work on the classes involved in the benchmark tests for new features or debugging issues in different environments.

**Comment by kinow** [comment]
Tried running the following to test the pull request:

```
$ mvn clean install
$ mvn test -Pbenchmark -e -X
```

And it kept failing with:

```
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 9.718 s
[INFO] Finished at: 2017-02-22T20:00:57+13:00
[INFO] Final Memory: 32M/486M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.5.0:exec (benchmark) on project commons-lang3: Command execution failed. Process exited with an error: 1 (Exit value: 1) -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.5.0:exec (benchmark) on project commons-lang3: Command execution failed.
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Command execution failed.
	at org.codehaus.mojo.exec.ExecMojo.execute(ExecMojo.java:302)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	... 20 more
Caused by: org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1)
	at org.apache.commons.exec.DefaultExecutor.executeInternal(DefaultExecutor.java:404)
	at org.apache.commons.exec.DefaultExecutor.execute(DefaultExecutor.java:166)
	at org.codehaus.mojo.exec.ExecMojo.executeCommandLine(ExecMojo.java:764)
	at org.codehaus.mojo.exec.ExecMojo.executeCommandLine(ExecMojo.java:711)
	at org.codehaus.mojo.exec.ExecMojo.execute(ExecMojo.java:289)
	... 22 more
```

Tried playing with dependency versions and settings, using Commons CSV as example (which works for me, as long as manually install one maven dependency as per pom.xml comment) with no luck.

Will give it another try tomorrow at work.

```
$ mvn -v
Apache Maven 3.3.9 (bb52d8502b132ec0a5a3f4c09453c07478323dc5; 2015-11-11T05:41:47+13:00)
Maven home: /opt/maven
Java version: 1.8.0_121, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-8-oracle/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "4.4.0-62-generic", arch: "amd64", family: "unix"
```

**Comment by C0rWin** [comment]
I'd guess it fails since there is no JMH tests available in the project, so it fails to execute them.

**Comment by C0rWin** [comment]
Actually once this one will be in, I have a few tests based on JMH which I can push.

**Comment by kinow** [comment]
@C0rWin what about trying to add the test from [LANG-1110](https://issues.apache.org/jira/browse/LANG-1110) to the pull request? Look at the attachments section.

**Comment by C0rWin** [comment]
I'm fine with it, will add the test from LANG-1110.

**Comment by PascalSchumacher** [comment]
Thanks!

**Pull request closed without merging**